Add installStrategy for addon and upgrade to v1beta1

qiujian16 opened the pull request:
Signed-off-by: Jian Qiu <jqiu@redhat.com>

qiujian16 committed:
Add installStrategy for addon and upgrade to v1beta1

Signed-off-by: Jian Qiu <jqiu@redhat.com>

haoqing0110 commented on addon/v1beta1/types_clustermanagementaddon.go:
```diff
@@ -0,0 +1,149 @@
+	Placement PlacementRef `json:",inline"`
+	// Configs is the configuration of managedClusterAddon during installation.
+	// User can override the configuration by updating the managedClusterAddon directly.
+	Configs []AddOnConfig `json:"configs,omitempty"`
```
Need `+optional` flag here?

haoqing0110 commented:
/lgtm

qiujian16 committed:
Add test cases for addon v1beta1 api

For now we need to set storage of mca to v1alpha1,
otherwise the installNamespace field will be lost

Signed-off-by: Jian Qiu <jqiu@redhat.com>

skeeey commented on addon/v1beta1/types_managedclusteraddon.go:
```diff
@@ -0,0 +1,279 @@
+type ManagedClusterAddOnSpec struct {
+	// configs is a list of add-on configurations.
+	// In scenario where the current add-on has its own configurations.
+	// An empty list means there are no defautl configurations for add-on.
```
An empty list means there are no owned configurations for add-on？

qiujian16 commented:
/hold
I want to rethink a bit upgrade path

zhujian7 commented on addon/v1beta1/types_managedclusteraddon.go:
```diff
@@ -0,0 +1,279 @@
+	// +kubebuilder:default=open-cluster-management-agent-addon
+	// +kubebuilder:validation:MaxLength=63
+	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+	RegistrationNamespace string `json:"registrationNamespace,omitempty"`
```
Not sure if using Namespace directly is better since it is in the RegistrationConfig struct.
